### PR TITLE
Remove reference tensor creation in producer tensor indexing path

### PR DESCRIPTION
**Pull request by shmsong**
Fixes #ISSUE_NUMBER

### CODE REVIEW

File: torch/csrc/jit/codegen/cuda/index_compute.cpp [hunk 1]
```diff
@@ -1241,37 +1334,68 @@ Val* hoistProducerIndex(
...
+  //   consumer: [b1, i2, i3]
+  //   producer: [i2, i3].
+  // Suppose the consumer is transformed as:
+  //   consumer: [(b1*i2)*i3]
+  // Then the producer would be transformed when indexed:
+  //   producer: [i2*i3]
+  // Assuming i2 and i3 are contiguous, the producer indexing is done
+  // with the mreged i2*i3 domain, but there's no domain in the
+  // cosumer that maps with the producer indexed domain.
+  // It seems non-trivial to support patterns like this. Skip for now.
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_compute.cpp:1356]: Yeah, I agree. I think this pattern isn't important to handle any time soon. More important I think we need to change what we consider for hoisting, as right now we only consider hoisting the final indexing at the root. Seems we would want a progressive hoisting where we check hoisting indexing closer to the leaf domains.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp:1356]: I don't see value here, but actually couldn't we just use best effort replay to ignore the broadcast axes in consumer? If we can map across while ignoring broadcast which we should be able to here, we should be able to hoist.

Comment (csarofeen) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp:1356]: Oh, just realized this comment was copy-pasted 😆

File: torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp [hunk 2]
```diff
@@ -151,6 +165,32 @@ IndexingParameters getGlobalIndexParameters(
       loop_indexing.loopDomains(),
       index_parameters.initial_concrete_id_index);
 
+  // Setup double buffer increment for producer case:
```
Comment (csarofeen) [inline torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp:168]: I don't remember the double buffering part of indexing. @naoyam could you double check the double buffering portions of the changes in this PR stack.

**Review by csarofeen — APPROVED** [review]
LGTM

File: torch/csrc/jit/codegen/cuda/index_compute.cpp (hunk 1 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/index_compute.cpp:1356]: We might have a chance at fixing this at later stage of refactoring. Will revisit.

File: torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp (hunk 2 above)
Comment (shmsong) [reply torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp:168]: This is just incrementing producer index by one on the main stage of double buffer loop, if the consumer is double buffered, i.e. prefetching. We can now make the logic more explicit. 

Will do that in a follow up as the merged PR is already getting huge.

Comment (naoyam) [reply torch/csrc/jit/codegen/cuda/index_idgraph_utils.cpp:168]: That's right. `double_buffer_loop` is a loop that's double buffered in the current loop nest. We can just skip the rest if it's nullptr.

**Pull request merged**